cmd/xgql/main.go

```diff
@@ -104,19 +108,44 @@ func main() {
 	log := logging.NewLogrLogger(zl.WithName("xgql"))
 
 	kingpin.FatalIfError(otelruntime.Start(), "cannot add OpenTelemetry runtime instrumentation")
+
+	// Observing a log line during startup as follows:
+	// 2021/06/08 10:03:26 <nil>
+	// This appears to be fixed with https://github.com/open-telemetry/opentelemetry-go/pull/1851
+	// However, there is no release for opentelemetry-go with this fix yet.
 	res := resource.NewWithAttributes(attribute.String("service.name", "crossplane.io/gql"))
 
 	// OpenTelemetry metrics.
 	prom, err := prometheus.InstallNewPipeline(prometheus.Config{}, basic.WithResource(res))
 	kingpin.FatalIfError(err, "cannot create OpenTelemetry Prometheus exporter")
 
-	// OpenTelemetry tracing.
-	if *agent != "" {
-		flush, err := jaeger.InstallNewPipeline(jaeger.WithAgentEndpoint(*agent), jaeger.WithSDKOptions(
-			trace.WithSampler(trace.ParentBased(trace.TraceIDRatioBased(*ratio))),
-			trace.WithResource(res)))
+	tpOpts := []trace.TracerProviderOption{
+		trace.WithSampler(trace.ParentBased(trace.TraceIDRatioBased(*ratio))),
+		trace.WithResource(res),
+	}
+	switch *tracer {
+	case "jaeger":
+		log.Debug("Enabling Jaeger tracer")
+		if *jaegerAgent == "" {
+			kingpin.Fatalf("cannot enable Jaeger tracing, jaeger agent not set")
+		}
+		host, port, err := net.SplitHostPort(*jaegerAgent)
+		kingpin.FatalIfError(err, "cannot parse jaeger agent as host:port")
+
+		exp, err := jaeger.NewRawExporter(jaeger.WithAgentEndpoint(jaeger.WithAgentHost(host), jaeger.WithAgentPort(port)))
 		kingpin.FatalIfError(err, "cannot create OpenTelemetry Jaeger exporter")
-		defer flush()
+		jp := trace.NewTracerProvider(append(tpOpts, trace.WithSyncer(exp))...)
+		defer func() {
+			err = jp.Shutdown(context.Background())
+			kingpin.FatalIfError(err, "cannot shutdown Jaeger exporter")
+		}()
+	case "gcp":
+		log.Debug("Enabling GCP tracer")
+		_, shutdown, err := texporter.InstallNewPipeline([]texporter.Option{}, tpOpts...)
+		kingpin.FatalIfError(err, "cannot create OpenTelemetry GCP exporter")
+		defer shutdown()
+	default:
+		log.Debug("Tracing not enabled")
```

Comment:
I'd slightly prefer to avoid this log line; I know it's debug but I'm not sure how useful logging that something is not enabled is. I would imagine the absence of a log line saying tracing is enabled would be sufficient.